File: csharp/1297.maximum-number-of-occurrences-of-a-substring.cs
/*
 * @lc app=leetcode id=1297 lang=csharp
 *
 * [1297] Maximum Number of Occurrences of a Substring
 */

using System.Collections.Generic;
using System.Linq;

// @lc code=start
public class Solution {
    public int MaxFreq(string s, int maxLetters, int minSize, int maxSize) {
        var count = new Dictionary<string, int>();
        int l = 0, charCount = 0, n = s.Length;
        int[] chars = new int[26];
        for (int r = 0; r < n; ++r) {
            int c = (int)s[r] - (int)'a';
            if (chars[c]++ == 0) {
                ++charCount;
            }
            if (r - l >= minSize) {
                int p = (int)s[l++] - (int)'a';
                if (--chars[p] == 0) {
                    --charCount;
                }
            }
            if (r - l + 1 == minSize && charCount <= maxLetters) {
                string sub = s.Substring(l, minSize);
                if (count.TryGetValue(sub, out int t)) {
                    count[sub] = t + 1;
                } else {
                    count[sub] = 1;
                }
            }
        }
        return count.Values.DefaultIfEmpty().Max();
    }
}
// @lc code=end

File: java/1297.maximum-number-of-occurrences-of-a-substring.java
/*
 * @lc app=leetcode id=1297 lang=java
 *
 * [1297] Maximum Number of Occurrences of a Substring
 */

import java.util.HashMap;
import java.util.Map;

// @lc code=start

class Solution {
    public int maxFreq(String s, int maxLetters, int minSize, int maxSize) {
        int n = s.length(), ret = 0, d = 0;
        int[] count = new int[26];
        Map<String, Integer> map = new HashMap<>();
        StringBuilder sb = new StringBuilder(minSize);
        for (int r = 0; r < n; ++r) {
            char c = s.charAt(r);
            if (count[c - 'a']++ == 0) {
                ++d;
            }
            if (r >= minSize) {
                if (count[s.charAt(r - minSize) - 'a']-- == 1) {
                    --d;
                }
                sb.deleteCharAt(0);
            }
            sb.append(c);
            if (r < minSize - 1 || d > maxLetters) {
                continue;
            }
            String k = sb.toString();
            int p = map.getOrDefault(k, 0) + 1;
            map.put(k, p);
            ret = Math.max(p, ret);
        }
        return ret;
    }
}
// @lc code=end

